src/core/execute.c

```diff
@@ -2446,8 +2447,12 @@ static int setup_keyring(
 
         /* And now, make the keyring owned by the service's user */
         if (uid_is_valid(uid) || gid_is_valid(gid))
-                if (keyctl(KEYCTL_CHOWN, keyring, uid, gid, 0) < 0)
-                        return log_unit_error_errno(u, errno, "Failed to change ownership of session keyring: %m");
+                if (keyctl(KEYCTL_CHOWN, keyring, uid, gid, 0) < 0) {
+                        log_unit_warning_errno(u, errno, "Failed to change ownership of session keyring: %m");
+                        /* well, the kernel didn't let us - clean up the invocation_id just in case */
+                        if (keyctl(KEYCTL_UNLINK, key, keyring, 0, 0) < 0)
+                                return log_unit_error_errno(u, errno, "Failed to unlink (clean-up) key, after failing to change ownership: %m");
+                }
```

Comment:
so, with this change we'll stil invoke unprivileged services with a very weird keyring environment: it will still have its own session keyring, but it won't be owned by it... I am not convinced we should do something like this: pass an incorrectly configured keyring to services

Reply:
I was not sure if it was save to unlink / delete session keyring. Also from the manpages I was not sure if creating a session keyring actually does that, or returns the user keyring. Manpage implies it is valid to simply return user keyring instead of the session keyring.